sshd-common/src/main/java/org/apache/sshd/common/mac/Poly1305Mac.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sshd.common.mac;
+
+import java.nio.BufferOverflowException;
+import java.security.InvalidKeyException;
+import java.util.Arrays;
+
+import org.apache.sshd.common.util.NumberUtils;
+import org.apache.sshd.common.util.buffer.BufferUtils;
+
+/**
+ * Poly1305 one-time message authentication code. This implementation is derived from the public domain C library
+ * <a href="https://github.com/floodyberry/poly1305-donna">poly1305-donna</a>.
+ *
+ * @see <a href="http://cr.yp.to/mac/poly1305-20050329.pdf">The Poly1305-AES message-authentication code</a>
+ */
+public class Poly1305Mac implements Mac {
+    private static final int BLOCK_SIZE = 16;
+
+    private int r0;
+    private int r1;
+    private int r2;
+    private int r3;
+    private int r4;
+    private int s1;
+    private int s2;
+    private int s3;
+    private int s4;
+    private int k0;
+    private int k1;
+    private int k2;
+    private int k3;
+
+    private int h0;
+    private int h1;
+    private int h2;
+    private int h3;
+    private int h4;
+    private final byte[] currentBlock = new byte[BLOCK_SIZE];
+    private int currentBlockOffset;
+
+    public Poly1305Mac() {
+        // empty
+    }
+
+    @Override
+    public String getAlgorithm() {
+        return "Poly1305";
+    }
+
+    @Override
+    public void init(byte[] key) throws Exception {
+        if (NumberUtils.length(key) != 32) {
```

Comment:
Maybe add a Poly1305Mac.KEY_BYTES constant for this 32, and use in ChachaEngine.polyKey()?